source/solvers/mf_navier_stokes_operators.cc

```diff
@@ -0,0 +1,810 @@
+
+  this->viscosity = viscosity;
+
+  this->compute_element_size();
+
+  constrained_indices.clear();
+  for (auto i : this->matrix_free.get_constrained_dofs())
+    constrained_indices.push_back(i);
+  constrained_values.resize(constrained_indices.size());
+
+  if (this->matrix_free.get_mg_level() != numbers::invalid_unsigned_int)
+    {
+      std::vector<types::global_dof_index> interface_indices;
+      IndexSet                             refinement_edge_indices;
+      refinement_edge_indices = get_refinement_edges(this->matrix_free);
+      refinement_edge_indices.fill_index_vector(interface_indices);
+
+      edge_constrained_indices.clear();
+      edge_constrained_indices.reserve(interface_indices.size());
+      edge_constrained_values.resize(interface_indices.size());
+      const IndexSet &locally_owned =
+        this->matrix_free.get_dof_handler().locally_owned_mg_dofs(
+          this->matrix_free.get_mg_level());
+      for (unsigned int i = 0; i < interface_indices.size(); ++i)
+        if (locally_owned.is_element(interface_indices[i]))
+          edge_constrained_indices.push_back(
+            locally_owned.index_within_set(interface_indices[i]));
+
+      this->has_edge_constrained_indices =
+        Utilities::MPI::max(edge_constrained_indices.size(),
+                            dof_handler.get_communicator()) > 0;
+
+      if (this->has_edge_constrained_indices)
+        {
+          edge_constrained_cell.resize(matrix_free.n_cell_batches(), false);
+
+          VectorType temp;
+          matrix_free.initialize_dof_vector(temp);
+
+          for (unsigned int i = 0; i < edge_constrained_indices.size(); ++i)
+            temp.local_element(edge_constrained_indices[i]) = 1.0;
+
+          temp.update_ghost_values();
+
+          FECellIntegrator integrator(matrix_free);
+
+          for (unsigned int cell = 0; cell < matrix_free.n_cell_batches();
+               ++cell)
+            {
+              integrator.reinit(cell);
+              integrator.read_dof_values(temp);
+
+              for (unsigned int i = 0; i < integrator.dofs_per_cell; ++i)
+                if ((integrator.begin_dof_values()[i] ==
+                     VectorizedArray<number>()) == false)
+                  {
+                    edge_constrained_cell[cell] = true;
+                    break;
+                  }
+            }
+
+          unsigned int count = 0;
+          for (const auto i : edge_constrained_cell)
+            if (i)
+              count++;
+
+          const unsigned int count_global =
+            Utilities::MPI::sum(count, dof_handler.get_communicator());
+
+          const unsigned int count_cells_global =
+            Utilities::MPI::sum(matrix_free.n_cell_batches(),
+                                dof_handler.get_communicator());
+
+          if (Utilities::MPI::this_mpi_process(
+                dof_handler.get_communicator()) == 0)
+            std::cout << count_global << " " << count_cells_global << std::endl;
```

Comment:
Shouldn't this be optional?

Reply:
At least if the output is not optional, it should be a bit more verbose in what it means

Reply:
Yes, it should be optional and I will add a comment explaining the purpose of it.